_posts/2020-06-18-Metal3-dev-env-BareMetal-Cluster-Deployment.md

```diff
@@ -0,0 +1,396 @@
+
+```console
+kubectl describe MachineDeployment ${CLUSTER_NAME} -n metal3
+```
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1alpha3
+kind: MachineDeployment
+metadata:
+  [....]
+  ownerReferences:
+  - apiVersion: cluster.x-k8s.io/v1alpha3
+    kind: Cluster
+    name: bmetalcluster
+    uid: aec0f73b-a068-4992-840d-6330bf943d22
+  resourceVersion: "66257"
+  selfLink: /apis/cluster.x-k8s.io/v1alpha3/namespaces/metal3/machinedeployments/bmetalcluster
+  uid: f598da43-0afe-44e4-b793-cd5244c13f4e
+spec:
+  clusterName: bmetalcluster
+  minReadySeconds: 0
+  progressDeadlineSeconds: 600
+  replicas: 1
+  revisionHistoryLimit: 1
+  selector:
+    matchLabels:
+      cluster.x-k8s.io/cluster-name: bmetalcluster
+      nodepool: nodepool-0
+  strategy:
+    rollingUpdate:
+      maxSurge: 1
+      maxUnavailable: 0
+    type: RollingUpdate
+  template:
+    metadata:
+      labels:
+        cluster.x-k8s.io/cluster-name: bmetalcluster
+        nodepool: nodepool-0
+    spec:
+      bootstrap:
+        configRef:
+          apiVersion: bootstrap.cluster.x-k8s.io/v1alpha3
+          kind: KubeadmConfigTemplate
+          name: bmetalcluster-workers
+      clusterName: bmetalcluster
+      infrastructureRef:
+        apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
+        kind: Metal3MachineTemplate
+        name: bmetalcluster-workers
+      version: v1.18.0
+status:
+  observedGeneration: 1
+  phase: ScalingUp
+  replicas: 1
+  selector: cluster.x-k8s.io/cluster-name=bmetalcluster,nodepool=nodepool-0
+  unavailableReplicas: 1
+  updatedReplicas: 1
+```
+
+<br/>
+To check the status we can follow steps similar to Controlplane case :
+
+```console
+kubectl get bmh -n metal3 -w
+```
+> We can see the live status of the node being provisioned. As mentioned before `bmh` is short representation for `BareMetalHosts`.
+
+```console
+kubectl get Machine -n metal3 -w
+```
+> This shows the status of Machines associated with workers, apart from the one for Controlplane, and we can watch the status of provisioning under PHASE 
+
+```console
+sudo virsh net-dhcp-leases baremetal
+```
+> To get the nodes IP
+
+```console
+ssh metal3@{master-node-ip}
+```
+> To check if its added to the cluster
+
+```console
+ssh metal3@{node-ip}
+```   
+> If you want to login to the node
+
+```console
+kubectl scale --replicas=3 MachineDeployment ${CLUSTER_NAME} -n metal3
+```
+> We can add or remove workers to the cluster, we can scale up the MachineDeployment up or down, in this example we are adding 2 more worker nodes, making the total nodes = 3
+
+<br/>
+<br/>
+
+### Deprovisioning
+
+All of the previous components have corresponding deprovisioning scripts which use config files, in the previously mentioned manifest directory, and use them to clean up worker, controlplane and cluster.
+For example if you wish to deprovision the cluster, you would do :
+
+```console
+sh ${metal3-dev-env-path}/scripts/v1alphaX/deprovision_cluster.sh
```

Comment:
I would recommend to also run the other deprovision scripts as some elements such as the metal3machinetemplates are not automatically deleted when you delete the cluster only. The rest should be though.

Reply:
I have included steps for other deprovision scripts and a note explaining it.